src/sources/fluent/message.rs

```diff
@@ -0,0 +1,347 @@
+                .unwrap_or_else(|| Value::Bytes(i.to_string().into())),
+            rmpv::Value::F32(f) => Value::Float(f.into()),
+            rmpv::Value::F64(f) => Value::Float(f),
+            rmpv::Value::String(s) => Value::Bytes(s.into_bytes().into()),
+            rmpv::Value::Binary(bytes) => Value::Bytes(bytes.into()),
+            rmpv::Value::Array(values) => Value::Array(
+                values
+                    .into_iter()
+                    .map(|value| Value::from(FluentValue(value)))
+                    .collect(),
+            ),
+            rmpv::Value::Map(values) => {
+                // Per
+                // https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#message-modes
+                // we should expect that keys are always stringy. Ultimately a
+                // lot hinges on what
+                // https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#grammar
+                // defines 'object' as.
+                //
+                // The current implementation will SILENTLY DROP non-stringy keys.
+                Value::Map(
+                    values
+                        .into_iter()
+                        .filter_map(|(key, value)| {
+                            key.as_str()
+                                .map(|k| (k.to_owned(), Value::from(FluentValue(value))))
+                        })
+                        .collect(),
+                )
+            }
+            rmpv::Value::Ext(code, bytes) => {
+                let mut fields = BTreeMap::new();
+                fields.insert(
+                    String::from("msgpack_extension_code"),
+                    Value::Integer(code.into()),
+                );
+                fields.insert(String::from("bytes"), Value::Bytes(bytes.into()));
+                Value::Map(fields)
+            }
+        }
+    }
+}
+
+/// Fluent message timestamp.
+///
+/// Message timestamps can be a unix timestamp or EventTime messagepack ext.
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(untagged)]
+pub(super) enum FluentTimestamp {
+    #[serde(with = "ts_seconds")]
+    Unix(DateTime<Utc>),
+    Ext(FluentEventTime),
+}
+
+impl From<FluentTimestamp> for Value {
+    fn from(timestamp: FluentTimestamp) -> Self {
+        match timestamp {
+            FluentTimestamp::Unix(timestamp) | FluentTimestamp::Ext(FluentEventTime(timestamp)) => {
+                Value::Timestamp(timestamp)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::sources::fluent::message::FluentValue;
+    use approx::assert_relative_eq;
+    use quickcheck::quickcheck;
+    use std::collections::BTreeMap;
+    use vector_core::event::Value;
+
+    quickcheck! {
+      fn from_bool(input: bool) -> () {
+          assert_eq!(Value::from(FluentValue(rmpv::Value::Boolean(input))),
+              Value::Boolean(input))
+        }
+    }
+
+    quickcheck! {
+      fn from_u64(input: i64) -> () {
```

Comment:
```suggestion
      fn from_u64(input: u64) -> () {
```

I think this one and the one below are reversed.

Reply:
Right you are. Corrected in aaf7bf4